src/core/utilities/generate_aliases.ts

```diff
@@ -1,3 +1,133 @@
+function cartesianProduct<T>(arr: T[][]): T[][] {
+  return arr.reduce(
+    (a, b) => {
+      return a
+        .map(x => {
+          return b.map(y => {
+            return x.concat(y);
+          });
+        })
+        .reduce((c, d) => c.concat(d), []);
+    },
+    [[]] as T[][]
+  );
+}
+
+export function* aliasesFromPattern(query: string) {
+  if (!query || query.length === 0) {
+    throw new Error('Tried to parse null or empty string');
+  }
+
+  let i = 0;
+  const currentChar = () => query[i];
+
+  function match(c: string) {
+    const char = currentChar();
+    if (c.includes(char)) {
+      i++;
+      return char;
+    }
+    else if (i >= query.length) {
+      throw new Error(
+        `Ran off the end of the string expecting to find: '${c}', i: ${i}`
+      );
+    } else {
+      throw new Error(
+        `Unexpected character: '${currentChar()}'. Expected '${c}'.`
+      );
+    }
+  }
+
+  //Intrinsics
+  const whitespace = () => match(' ');
+  const letter = () =>
+    match(
+      'ABCDEFGHIJKLMNOPQRSTUVWXYZ' +
+      'abcdefghijklmnopqrstuvwxyz' +
+      'ÁáÍíøOóÚúÜüÑñ' +
+      ' '
+    );
+  const token = () => [kleenePlus(letter).join('')];
+  const constant = (s: string) => {
+    for (let i = 0; i < s.length; i++) match(s.charAt(i));
+  };
+  const separator = () => match('|,;');
+
+  //Lists
+  const listGap = () => {
+    separator();
+    kleene(whitespace);
+  };
+  const group = () => kleenePlus(phrase, listGap);
+
+  //Branches
+  function choice(): string[] {
+    constant('(');
+    const contents = group();
+    constant(')');
+    return ([] as string[]).concat(...contents);
+  }
+
+  function optional(): string[] {
+    constant('[');
+    const contents = group();
+    constant(']');
+    const result = ([] as string[]).concat(...contents);
+    result.push('');
+    return result;
+  }
+
+  const term = () => union(choice, optional, token);
+  const phrase = () => cartesianProduct(kleenePlus(term)).map(a => a.join(''));
```

Comment:
Haha, took a second to decompose this one.

`kleenePlus(...)` generates a list of branches (where each branch is a list of tokens), then the Cartesian enumerates all paths through these branches. 

Wouldn't be a bad idea to document this! A little explanation by the dev saves a lot of reverse engineering by the reader.